### PR TITLE
feat(backup): store storage class name to backup volume

### PR DESCRIPTION
**Pull request by ChanYiLin**
ref: [longhorn/longhorn#3506](https://github.com/longhorn/longhorn/issues/3506)

blocked by: 
- backupstore: https://github.com/longhorn/backupstore/pull/133
- longhorn-engine: https://github.com/longhorn/longhorn-engine/pull/904
- longhorn-instance-manager: https://github.com/longhorn/longhorn-instance-manager/pull/230

### CODE REVIEW

**Comment by ChanYiLin** [comment]
> This pull request is now in conflicts. Could you fix it @ChanYiLin? 🙏

conflicts resolved

**Comment by ChanYiLin** [comment]
> @ChanYiLin Resolve the conflicts and check failure.

Hi @innobead , dependencies has been updated and the conflicts has been resolved, thanks!

File: manager/kubernetes.go [hunk 1]
```diff
@@ -34,9 +34,17 @@ func (m *VolumeManager) PVCreate(name, pvName, fsType, secretNamespace, secretNa
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to get longhorn static storage class name for PV %v creation: %v", pvName, err)
+		}
+		if backupVolumeName, isExist := v.Labels[types.LonghornLabelBackupVolume]; isExist && backupVolumeName != "" {
+			backupVolume, _ := m.ds.GetBackupVolumeRO(backupVolumeName)
+			if backupVolume != nil {
+				storageClassName = backupVolume.Status.StorageClassName
+			}
+		}
 	}
```
Comment (innobead) [inline manager/kubernetes.go:48]: Can do it reverse, also make it clear to understand how storaglcass is determined. 

```go
if storageClassName == "" {
    if backupVolumeName := v.Labels[types.LonghornLabelBackupVolume]; backupVolumeName != "" {
        backupVolume, _ := m.ds.GetBackupVolumeRO(backupVolumeName)
        if backupVolume != nil {
            storageClassName = backupVolume.Status.StorageClassName
        }
    }
}

if storageClassName == "" {
    storageClassName, err = m.ds.GetSettingValueExisted(types.SettingNameDefaultLonghornStaticStorageClass)
    if err != nil {
    return nil, fmt.Errorf("failed to get longhorn default static storage class name for PV %v creation: %v", pvName, err)
    }
}

```

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Some comments need to resolve.

File: controller/backup_controller.go [hunk 2]
```diff
@@ -642,6 +642,19 @@ func (bc *BackupController) checkMonitor(backup *longhorn.Backup, volume *longho
 		return nil, err
 	}
 
+	// get storage class of the pvc binding with the volume
+	kubernetesStatus := &volume.Status.KubernetesStatus
+	storageClassName := ""
+	if kubernetesStatus.PVCName != "" && kubernetesStatus.LastPVCRefAt == "" {
+		pvc, _ := bc.ds.GetPersistentVolumeClaim(kubernetesStatus.Namespace, kubernetesStatus.PVCName)
+		logrus.Infof("====[DEBUG]==== Namespace: %v PVCName: %v", kubernetesStatus.Namespace, kubernetesStatus.PVCName)
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply controller/backup_controller.go]: removed, thanks

File: k8s/pkg/apis/longhorn/v1beta2/backupvolume.go [hunk 3]
```diff
@@ -48,6 +48,9 @@ type BackupVolumeStatus struct {
 	// the backing image checksum.
 	// +optional
 	BackingImageChecksum string `json:"backingImageChecksum"`
+	// the storage class name of pv/pvc binding with the volume.
+	// +optional
+	StorageClassName string `json:"storageClassName"`
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply k8s/pkg/apis/longhorn/v1beta2/backupvolume.go:53]: already created the PR, thanks
https://github.com/longhorn/longhorn/pull/5939/files

[the remark this answers is not in the document]
Comment (innobead) [reply k8s/pkg/apis/longhorn/v1beta2/backupvolume.go:53]: also, need https://github.com/longhorn/longhorn-manager/blob/master/k8s/crds.yaml. (in the future, we will make both sync..)

File: go.mod [hunk 4]
```diff
@@ -45,10 +45,10 @@ require (
 	github.com/jinzhu/copier v0.3.5
 	github.com/kubernetes-csi/csi-lib-utils v0.6.1
 	github.com/longhorn/backing-image-manager v1.4.0-rc1.0.20230521151917-38ff27cc2cbb
-	github.com/longhorn/backupstore v0.0.0-20230518134155-aa42d35366e3
+	github.com/longhorn/backupstore v0.0.0-20230522064032-228beb21aa3c
 	github.com/longhorn/go-iscsi-helper v0.0.0-20230425064248-72f136f48524
-	github.com/longhorn/longhorn-engine v1.4.0-rc1.0.20230518180217-e9f40c98d84c
-	github.com/longhorn/longhorn-instance-manager v1.4.0-rc1.0.20230519015737-23af7f034c53
+	github.com/longhorn/longhorn-engine v1.4.0-rc1.0.20230522141009-4f6770982ebd
```
[the remark this answers is not in the document]
Comment (innobead) [reply go.mod:50]: The merge history is quite weird, it should be rebased :thinking:

File: manager/kubernetes.go (hunk 1 above)
Comment (ChanYiLin) [reply manager/kubernetes.go]: Thanks, this looks better!

File: k8s/pkg/apis/longhorn/v1beta2/backupvolume.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply k8s/pkg/apis/longhorn/v1beta2/backupvolume.go:53]: updated, thanks

**Comment by ChanYiLin** [comment]
Hi @innobead ,fixed the comments, thanks!

**Review by innobead — APPROVED** [review]

**Pull request merged**